src/combinatorics/hindman.lean

```diff
@@ -0,0 +1,143 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import topology.stone_cech
+import data.stream
+
+/-!
+# Hindman's theorem on finite sums
+
+We prove Hindman's theorem on finite sums, using idempotent ultrafilters.
+
+Given an infinite sequence `a₀, a₁, a₂, …` of positive integers, the set `FS(a₀, …)` is the set
+of positive integers that can be expressed as a finite sum of `aᵢ`'s, without repetition. Hindman's
+theorem asserts that whenever the positive integers are finitely colored, there exists a sequence
+`a₀, a₁, a₂, …` such that `FS(a₀, …)` is monochromatic. We prove the theorem for a general
+semigroup `M` instead of `ℕ+` since it is no harder, although this special case implies the general
+case.
+
+The idea of the proof is to extend the addition `(+) : M → M → M` to addition `(+) : βM → βM → βM`
+on the space `βM` of ultrafilters on `M`. One can prove that if `U` is an _idempotent_ ultrafilter,
+i.e. `U + U = U`, then any `U`-large subset of `M` contains some set `FS(a₀, …)` (see
+`exists_FS_of_large`). And a general topological argument shows the existence of an idempotent
+ultrafilter (see `exists_idempotent`).
+
+## Main results
+
+- `exists_FS_of_finite_cover`: Hindman's theorem on finite sums.
+
+-/
+
+open filter
+
+instance {M} [has_add M] : has_add (ultrafilter M) := { add := λ U V, (+) <$> U <*> V }
+
+lemma eventually_add {M} [has_add M] (U V : ultrafilter M) (p : M → Prop) :
+  (∀ᶠ m in ↑(U + V), p m) ↔ ∀ᶠ m in U, ∀ᶠ m' in V, p (m + m') := iff.refl _
+
+instance {M} [add_semigroup M] : add_semigroup (ultrafilter M) :=
+{ add_assoc := λ U V W, ultrafilter.coe_inj.mp $ filter.ext' $ λ p,
+    by simp only [eventually_add, add_assoc]
+  ..ultrafilter.has_add }
+
+lemma left_continuity {M} [add_semigroup M] (V : ultrafilter M) : continuous (+ V) :=
+topological_space.is_topological_basis.continuous ultrafilter_basis_is_basis _ $
+set.forall_range_iff.mpr $ λ s, ultrafilter_is_open_basic { m : M | ∀ᶠ m' in V, m + m' ∈ s }
+
+lemma exists_idempotent (M : Type*) [add_semigroup M] [nonempty M] [topological_space M]
```

Comment:
This can probably also move to one of the first files that merges algebra and topology.

Reply:
I agree it makes sense to move this out, but couldn't find a good match among the existing files so I made a new one.